Draft pulumi-cdk 1.0 blog (#13434)
pulumi/docs: content/blog/aws-cdk-on-pulumi-1.0/index.md

---

> title: "Anouncing the 1.0 release of AWS CDK on Pulumi"

Author: I don't want to call it out in the title, but you're right that we should point this out early on so we don't get hopes up only to dash them at the end of the post

date: 2024-12-02T08:00:00-07:00
draft: true
meta_desc: "Enhanced support of AWS CDK constructs from within Pulumi. Combind Pulumi and AWS CDK resources and leverage the power, speed and verisitlity of the Pulumi Cloud Platform with CDK"
meta_image: todo.png
authors:
    - matt-jeffryes
    - cory-hall
    - florian-stadler
    - anton-tayanovskyy
tags:
    - aws-cdk
---

At Pulumi, we're committed to delivering the widest range of cloud infrastructure building blocks for use in your cloud engineering projects. 
In 2022, we introduced [preview](https://www.pulumi.com/blog/aws-cdk-on-pulumi/) support for integrating AWS CDK constructs into Pulumi programs 

> and today we're happy to announce the 1.0 release of our pulumi-cdk library. This first stable version completes support for common CDK features 

Author: ```suggestion
and today we're happy to announce the 1.0 release of our pulumi-cdk library for typescript. This first stable version completes support for common CDK features 
```

    url: "2ez3iazupm.us-west-2.awsapprunner.com"

Resources:
    + 4 created
```

And curl the endpoint:
```console
> curl https://$(pulumi stack output url)

   ______                             __        __      __  _                  __
  / ____/___  ____  ____ __________ _/ /___  __/ /___ _/ /_(_)___  ____  _____/ /
 / /   / __ \/ __ \/ __ `/ ___/ __ `/ __/ / / / / __ `/ __/ / __ \/ __ \/ ___/ /
/ /___/ /_/ / / / / /_/ / /  / /_/ / /_/ /_/ / / /_/ / /_/ / /_/ / / / (__  )_/
\____/\____/_/ /_/\__, /_/   \__,_/\__/\__,_/_/\__,_/\__/_/\____/_/ /_/____(_)
                 /____/

        Congratulations, your service has successfully deployed on AWS App Runner.

Open it in your browser at https://2ez3iazupm.us-west-2.awsapprunner.com/

Try the workshop at https://apprunnerworkshop.com
Read the docs at https://docs.aws.amazon.com/apprunner
```

## Pulumi takes CDK apps to the next level

Combine CDK constructs with any of the features of Pulumi programs to deploy faster, easier and to any cloud. Use Pulumi functions and stack references to connect to pre-existing infrastructure and mix in resources from any provider to bring *all* of your infrastructure under management. 

> For example, we can use a Pulumi function to look up AMIs to use when creating an EC2 instance in CDK. Then, we connect a cloudfront distribution over the public IP to provide caching and DDOS protection. 

Reviewer: ```suggestion
For example, we can use a Pulumi function to look up AMIs for creating an EC2 instance. Then, we connect a CloudFront distribution over the public IP to provide caching and DDOS protection. 
```

        vpc,
        instanceType: ec2.InstanceType.of(ec2.InstanceClass.T4G, ec2.InstanceSize.MICRO),
        machineImage,
    });

   // Return the public IP of the instance
   this.publicIp = this.asOutput(ec2Instance.publicIp);
});

// Pulumi's abstraction to create a CloudFront distribution
const distribution = new aws.cloudfront.Distribution("my-cdn-distribution", {
    enabled: true,
    origins: [{
        originId: ec2Instance.urn,
        domainName: ec2PublicIp.apply(ip => `${app.publicIp}.compute-1.amazonaws.com`), // Constructed domain name
        customOriginConfig: {
            originProtocolPolicy: "http-only",
            httpPort: 80,
            httpsPort: 80,
            originSslProtocols: ["TLSv1.2"],
        },
    }],
    defaultCacheBehavior: {
        targetOriginId: ec2Instance.urn,
        viewerProtocolPolicy: "allow-all",
        allowedMethods: ["GET", "HEAD", "OPTIONS"],
        cachedMethods: ["GET", "HEAD"],
        // Define cache behavior settings...
    },
    // Additional settings like price class, custom error responses, etc.
    // ...
    viewerCertificate: {
        cloudfrontDefaultCertificate: true,
    },

> });

Author: Cory, I would appreciate your sanity check on this I smashed some examples together based on our conversation Friday about the most compelling features from a CDK users perspective, but it probably needs some fixes.

Author: 🤦 I had intended to use Cloudflare, but was moving quickly and asked the AI the wrong question. We do have an example creating a record on cloudflare so I think we could test with that account+domain: [pulumi-cloudflare/examples/record/ts/index.ts at master · pulumi/pulumi-cloudflare](https://github.com/pulumi/pulumi-cloudflare/blob/master/examples/record/ts/index.ts)

import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as ecs_patterns from 'aws-cdk-lib/aws-ecs-patterns';

const app = new pulumicdk.App('app', (scope: pulumicdk.App) => {
    const stack = new pulumicdk.Stack('example-stack');

    const vpc = new ec2.Vpc(stack, 'MyVpc');
    const cluster = new ecs.Cluster(stack, 'fargate-service-autoscaling', { vpc });

    // Create Fargate Service
    const fargateService = new ecs_patterns.NetworkLoadBalancedFargateService(this, 'sample-app', {
        cluster,
        taskImageOptions: {
            image: ecs.ContainerImage.fromAsset(path.join(__dirname, './'), {
                // assetName is required and is used in the name of the ecr repository that is created
                assetName: 'cdk-fargate-example',
            }),
        },
    });
});
```

## Future direction

Currently AWS CDK on Pulumi is supported only for TypeScript users, but we're eager to bring the benefits of the CDK ecosystem to all Pulumi languages in the future. In the meantime, AWS CDK on Pulumi can be used within Component Packages implemented in TypeScript, and exposed to any Pulumi language.

> We're also exploring options for importing state from CDK applications that are already deployed with CloudFormation to enable seamless migration for existing CDK users. Please let us know if this is something you would be interested in in testing with us. 

Reviewer 2: I was talking with @t0yv0 and thought it might be cool to publish the tool as experimental alpha so at least we can point users somewhere.

Author: I would support that. based on who was interested in CDK support, I imagine this migration tool is going to be in high demand

    url: "2ez3iazupm.us-west-2.awsapprunner.com"

Resources:
    + 4 created
```

And curl the endpoint:

```console
> curl https://$(pulumi stack output url)

   ______                             __        __      __  _                  __
  / ____/___  ____  ____ __________ _/ /___  __/ /___ _/ /_(_)___  ____  _____/ /
 / /   / __ \/ __ \/ __ `/ ___/ __ `/ __/ / / / / __ `/ __/ / __ \/ __ \/ ___/ /
/ /___/ /_/ / / / / /_/ / /  / /_/ / /_/ /_/ / / /_/ / /_/ / /_/ / / / (__  )_/
\____/\____/_/ /_/\__, /_/   \__,_/\__/\__,_/_/\__,_/\__/_/\____/_/ /_/____(_)
                 /____/

        Congratulations, your service has successfully deployed on AWS App Runner.

Open it in your browser at https://2ez3iazupm.us-west-2.awsapprunner.com/

Try the workshop at https://apprunnerworkshop.com
Read the docs at https://docs.aws.amazon.com/apprunner
```

## Pulumi takes CDK apps to the next level

Combine CDK constructs with any of the features of Pulumi programs to deploy faster, easier and to any cloud. Use Pulumi functions and stack references to connect to pre-existing infrastructure and mix in resources from any provider to bring *all* of your infrastructure under management.

> For example, we can use CDK's `ecs_patterns` to quickly create a loadbalanced Fargate service, but route it with a record in an exisiting CloudFlare DNS zone.

Reviewer 3: `route it` reads a little weird to me, is `routing` a verb that can be done to a service? Perhaps `route to it` or `route traffic to it`

Author: ```suggestion
For example, we can use CDK's `ecs_patterns` to quickly create a loadbalanced Fargate service, but route traffic to it with a record in an exisiting CloudFlare DNS zone.
```

```ts
import * as pulumicdk from '@pulumi/cdk';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { AwsSolutionsChecks } from 'cdk-nag';

const app = new pulumicdk.App('app', (scope: pulumicdk.App): pulumicdk.AppOutputs => {
    const stack = new pulumicdk.Stack('example-stack');
    Aspects.of(stack).add(new AwsSolutionsChecks({ verbose: true }));

    new s3.Bucket(this, 'bucket');
});
```

#### Example Output

```console
[Error at /test-stack/bucket/Resource] AwsSolutions-S1: The S3 Bucket has server access logs disabled. The bucket should have server access logging enabled to provide detailed records for the requests that are made to the bucket.
[Error at /test-stack/bucket/Resource] AwsSolutions-S10: The S3 Bucket or bucket policy does not require requests to use SSL. You can use HTTPS (TLS) to help prevent potential attackers from eavesdropping on or manipulating network traffic using person-in-the-middle or similar attacks. You should allow only encrypted connections over HTTPS (TLS) using the aws:SecureTransport condition on Amazon S3 bucket policies.
```

### CDK assets

> CDK's assets make it super easy to bundle code and other files your application needs into an S3 bucket or a Docker container, eliminating extra steps from your build pipeline. For example you can use Docker container asset to deploy an Fargate service from local directory. (CDK will build the container image and push to ECR on `pulumi up`.):

Author: ```suggestion
CDK's assets make it easy to bundle code and other files your application needs into an S3 bucket or a Docker container, eliminating extra steps from your build pipeline. For example you can use Docker container asset to deploy a Fargate service from local directory. CDK will build the container image and push to ECR on `pulumi up`.
```

```ts
import * as pulumicdk from '@pulumi/cdk';
import * as aws from '@pulumi/aws';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as ecs_patterns from 'aws-cdk-lib/aws-ecs-patterns';

const app = new pulumicdk.App('app', (scope: pulumicdk.App) => {
    const stack = new pulumicdk.Stack('example-stack');

    const vpc = new ec2.Vpc(stack, 'MyVpc');
    const cluster = new ecs.Cluster(stack, 'fargate-service-autoscaling', { vpc });

    // Create Fargate Service
    const fargateService = new ecs_patterns.NetworkLoadBalancedFargateService(this, 'sample-app', {
        cluster,
        taskImageOptions: {
            image: ecs.ContainerImage.fromAsset(path.join(__dirname, './'), {
                // assetName is required and is used in the name of the ecr repository that is created
                assetName: 'cdk-fargate-example',
            }),
        },
    });
});
```

## Future direction

> Currently AWS CDK on Pulumi is supported only for TypeScript users, but we're eager to bring the benefits of the CDK ecosystem to all Pulumi languages in the future. In the meantime, AWS CDK on Pulumi can be used within Component Packages implemented in TypeScript, and exposed to any Pulumi language.

Reviewer 3: This could be made a little shorter. It is double-stating.

Author: ```suggestion
Currently AWS CDK on Pulumi is supported only for TypeScript users, but we're eager to bring the benefits of the CDK ecosystem to all Pulumi languages in the future. You can comment and upvote on [this tracking issue](https://github.com/pulumi/pulumi-cdk/issues/49), to let us know this is important to you. 
```

> We're also exploring options for importing state from CDK applications that are already deployed with CloudFormation to enable seamless migration for existing CDK users. Please let us know if this is something you would be interested in in testing with us.

Reviewer 3: This could use a contact link, e.g. "upvote this issue" or how to reach us with this.

---
title: "Anouncing the 1.0 release of AWS CDK on Pulumi"
date: 2024-12-02T08:00:00-07:00

> draft: true

Author: ```suggestion
```
